### PR TITLE
UNR-4877: Adapted spatial debugger to use FSubView instead of StaticComponentView

### PR DESCRIPTION
**Pull request by improbable-dmitrii**
#### Description
Split SpatialDebugger into two classes: `FSpatialDebuggerSystem` that interacts with Spatial using a SubView, and `ASpatialDebugger` that interacts with BPs, handles debug drawing and replicating worker regions.

#### Tests
Manual tests, ran around in existing maps in TestGym.

#### Documentation
Added a release note; this change doesn't change external API.

#### Primary reviewers
@jacquese @m-samiec were interested in seeing this PR.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h [hunk 1]
```diff
@@ -35,7 +36,45 @@ DECLARE_CYCLE_STAT(TEXT("Projection"), STAT_Projection, STATGROUP_SpatialDebugge
...
 DECLARE_CYCLE_STAT(TEXT("BuildText"), STAT_BuildText, STATGROUP_SpatialDebugger);
-DECLARE_CYCLE_STAT(TEXT("SortingActors"), STAT_SortingActors, STATGROUP_SpatialDebugger);
+
+namespace SpatialGDK
+{
+class FSubView;
+struct SpatialDebugging;
+} // namespace SpatialGDK
+
+class FSpatialDebuggerSystem
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:46]: 1. `FSpatialDebuggerSystem` is in the same vein as `ActorSystem` but sounds funny to me, especially just next to `ASpatialDebugger`.
1. Other Spatial-interacting classes are in `SpatialGDK` namespace and don't follow Unreal's `F`-prefixed naming for non-UCLASSes. Should I go for this naming here as well?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 2]
```diff
@@ -193,6 +194,7 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	FFilterPredicate ActorFilter;
 	TArray<FDispatcherRefreshCallback> ActorRefreshCallbacks;
 
+	TUniquePtr<FSpatialDebuggerSystem> SpatialDebuggerSystem;
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h:197]: I'm in two minds on TUniquePtr usage for the GDK - why we use raw pointers rather than private `TSharedPtr` in the NetDriver with `TWeakPtr` Get() functions?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 3]
```diff
@@ -446,6 +446,12 @@ void USpatialNetDriver::CreateAndInitializeCoreClasses()
 		MakeUnique<SpatialGDK::ActorSystem>(ActorNonAuthSubview, TombstoneActorSubview, this, &TimerManager, Connection->GetEventTracer());
 	ClientConnectionManager = MakeUnique<SpatialGDK::ClientConnectionManager>(SystemEntitySubview, this);
 
+	const SpatialGDK::FSubView& DebuggerSubView = Connection->GetCoordinator().CreateSubView(
+		IsServer() ? SpatialConstants::ACTOR_AUTH_TAG_COMPONENT_ID : SpatialConstants::SPATIAL_DEBUGGING_COMPONENT_ID,
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:450]: DEBUGGING component isn't a tag, any advantage of using it over ACTOR_NON_AUTH_TAG_COMPONENT_ID ?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:46]: 1. I agree it's a bit of a mouthful, but fairly explanatory, so I'm fine with it
2. Yep, no need to append the F/U/A etc. except when following Unreal's coding convention to detail UClass/UStruct and class hierarchy.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 4]
```diff
@@ -27,10 +27,174 @@
...
+	{
+		if (SubView->HasAuthority(EntityId, SpatialConstants::SERVER_AUTH_COMPONENT_SET_ID))
+		{
+			ActorAuthorityGained(EntityId);
+		}
+
+		return;
+	}
+
+	TWeakObjectPtr<AActor>* ExistingActor = EntityActorMapping.Find(EntityId);
```
Comment (jacquese) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:101]: why do we need this check? Is there a reason the subview wouldn't only give you adds for entities that don't currently exist in view?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 5]
```diff
@@ -446,6 +446,12 @@ void USpatialNetDriver::CreateAndInitializeCoreClasses()
 		MakeUnique<SpatialGDK::ActorSystem>(ActorNonAuthSubview, TombstoneActorSubview, this, &TimerManager, Connection->GetEventTracer());
 	ClientConnectionManager = MakeUnique<SpatialGDK::ClientConnectionManager>(SystemEntitySubview, this);
 
+	const SpatialGDK::FSubView& DebuggerSubView = Connection->GetCoordinator().CreateSubView(
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:449]: Rather than doing this always, I think it'd be nicer if this was only done on demand when the debugger is spawned. The spawning flow is a little different, servers spawn it via the gsm, and clients have it replicated over the wire. But seems like a nice use of a subview that it's only setup when it's actually needed.

**Review by jacquese — COMMENTED** [review]
first pass, generally a fan of the split between system and rest of debugger

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 6]
```diff
@@ -643,10 +643,7 @@ void USpatialSender::SendAuthorityIntentUpdate(const AActor& Actor, VirtualWorke
 	// This should always happen with USLB.
 	NetDriver->LoadBalanceEnforcer->ShortCircuitMaybeRefreshAuthorityDelegation(EntityId);
 
-	if (NetDriver->SpatialDebugger != nullptr)
-	{
-		NetDriver->SpatialDebugger->ActorAuthorityIntentChanged(EntityId, NewAuthoritativeVirtualWorkerId);
-	}
+	NetDriver->SpatialDebuggerSystem->ActorAuthorityIntentChanged(EntityId, NewAuthoritativeVirtualWorkerId);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:646]: check against null here

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 7]
```diff
@@ -27,10 +27,174 @@
 #include "Modules/ModuleManager.h"
 #include "Net/UnrealNetwork.h"
 
+#if UE_EDITOR
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:30]: Why is this needed?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h (hunk 1 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:46]: 2. Yeah, this is what I've been getting at - https://docs.unrealengine.com/en-US/ProductionPipelines/DevelopmentSetup/CodingStandard/#namingconventions details 
> Type names are prefixed with an additional upper-case letter
and
> Most other classes are prefixed by F
They don't really use classes without prefixes AFAIK.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 8]
```diff
@@ -27,10 +27,174 @@
...
+		}
+	}
+}
+
+void FSpatialDebuggerSystem::OnEntityAdded(const Worker_EntityId EntityId)
+{
+	check(NetDriver != nullptr);
+	if (NetDriver->IsServer())
+	{
+		if (SubView->HasAuthority(EntityId, SpatialConstants::SERVER_AUTH_COMPONENT_SET_ID))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:93]: Isn't this redundant as we handle it via the auth-gained in the entity delta above?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 9]
```diff
@@ -27,10 +27,174 @@
...
 
 DEFINE_LOG_CATEGORY(LogSpatialDebugger);
 
+FSpatialDebuggerSystem::FSpatialDebuggerSystem(USpatialNetDriver* InNetDriver, const FSubView& InSubView)
+	: NetDriver(InNetDriver)
+	, SubView(&InSubView)
+{
+	check(IsValid(InNetDriver));
+
+	EntityActorMapping.Reserve(ENTITY_ACTOR_MAP_RESERVATION_COUNT);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:44]: nit: This container is only used on clients, no need to reserve or iterate on servers

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 6 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:646]: Doesn't seem to be possible as it's a `TUniquePtr` initialized in `CreateAndInitializeCoreClasses`; do you think a `check(NetDriver->SpatialDebuggerSystem.IsValid())` would work?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 7 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:30]: There is a `GEditor` usage in this file, and it's defined in `Editor.h`. Just fixing a possible build issue.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 8 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:93]: No, in my experience, if there is an entity `ADD`, and you get it with authority, you won't get a `AuthorityGained` in `EntityDelta`. Will double check and come back.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h (hunk 1 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:46]: It does seem to be a convention in the GDK code though - should I move it to `SpatialGDK` and rename `SpatialDebuggerSystem`?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 3 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:450]: No need for a filter I guess?.. @jacquese has a similar point here: https://github.com/spatialos/UnrealGDK/pull/2869#discussion_r568887462
The first option (specifying `ACTOR_NON_AUTH_TAG_COMPONENT_ID` + a filter on clients) seems better to me, but seems to achieve the very same thing as using `SPATIAL_DEBUGGING_COMPONENT_ID` on clients with more involvement by querying more entities and them filtering them. Idk.

[the remark this answers is not in the document]
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:450]: Answered a similar point by @m-samiec here https://github.com/spatialos/UnrealGDK/pull/2869#discussion_r569319122

**Comment by m-samiec** [comment]
Looks good generally, can you also try spinning up some late connecting clients (via the LaunchClient.bat) and validate everything still all looks good? Also worth playing around with the debugger options if you haven't already, actor selection/deselection etc.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 3 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:450]: there won't be more entities queried necessarily, as the subview doesn't do any querying of its own. It potentially complicates the logic for an entity becoming complete, with the upside that if we want to change this component/add new components to this system we can do so without worrying about affecting this subview's entities, and also have it explicit that this subview cares about actors you are interested in.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 8 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:93]: yep, that's correct- adds are just markers, and you need to iterate through the main view for the entity you are told about to get the current state.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 5 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:449]: I did initially create it on `ASpatialDebugger::BeginPlay` but moved it here as per @jacquese's suggestion. `Advance` was called on `ASpatialDebugger::Tick` so the subview and the debugger system themselves were contained there; could move it to the NetDriver but this would spread the code across multiple moving parts which could be unnecessarily confusing.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 3 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Done!

**Comment by improbable-dmitrii** [comment]
Thanks for the note - I've ran around with `LaunchClient` clients and everything seems to work.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 7 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:30]: I think we normally use WITH_EDITOR here, but I'm not actually sure why/how they're different

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 10]
```diff
@@ -79,22 +247,65 @@ void ASpatialDebugger::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& Out
...
 
-void ASpatialDebugger::Tick(float DeltaSeconds)
+void ASpatialDebugger::BeginPlay()
 {
-	Super::Tick(DeltaSeconds);
+	Super::BeginPlay();
 
 	check(NetDriver != nullptr);
 
+	if (ensureMsgf(!NetDriver->SpatialDebuggerSystem.IsValid(), TEXT("SpatialDebugger system is already created at BeginPlay")))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:256]: I think this logic should all be on the netdriver, but trigger from the spatial debugger. You could also wrap the custom logic that's done in the constructor for clients in that functionality as well. And this call should be idempotent, because although we don't want to have more than a single spatial debugger in a world, there's current bugs where that can happen.

**Review by jacquese — APPROVED** [review]
lgtm % mike's comment on moving subview creation to a function in the netdriver

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 11]
```diff
@@ -79,22 +247,65 @@ void ASpatialDebugger::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& Out
...
+	{
+		const FSubView& DebuggerSubView = NetDriver->Connection->GetCoordinator().CreateSubView(
+			NetDriver->IsServer() ? SpatialConstants::ACTOR_AUTH_TAG_COMPONENT_ID : SpatialConstants::ACTOR_NON_AUTH_TAG_COMPONENT_ID,
+			[](const Worker_EntityId, const EntityViewElement& Element) -> bool {
+				return Element.Components.ContainsByPredicate(ComponentIdEquality{ SpatialConstants::SPATIAL_DEBUGGING_COMPONENT_ID });
```
Comment (jacquese) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:261]: Should this filter also include an `IsClient()` somewhere? Isn't it correct for the debugging system on servers to see the entity before it has a debugging component?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 12]
```diff
@@ -2815,17 +2800,50 @@ bool USpatialNetDriver::HasTimedOut(const float Interval, uint64& TimeStamp)
...
-void USpatialNetDriver::SetSpatialDebugger(ASpatialDebugger* InSpatialDebugger)
+void USpatialNetDriver::RegisterSpatialDebugger(ASpatialDebugger* InSpatialDebugger)
 {
-	check(!IsServer());
-	if (SpatialDebugger != nullptr)
+	if (!SpatialDebuggerSystem.IsValid())
 	{
-		UE_LOG(LogSpatialOSNetDriver, Error, TEXT("SpatialDebugger should only be set once on each client!"));
-		return;
+		typedef SpatialGDK::FSubView FSubView;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2807]: nit: these typdefs are a little unorthodox for gdk code, we normally as using statements

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 13]
```diff
@@ -107,55 +299,41 @@ void ASpatialDebugger::Tick(float DeltaSeconds)
...
+				.bConsumeInput = false;
+			LocalPlayerController->InputComponent->BindKey(HighlightActorKey, IE_Pressed, this, &ASpatialDebugger::OnHighlightActor)
+				.bConsumeInput = false;
 		}
+	}
+}
 
-		FontRenderInfo.bClipText = true;
-		FontRenderInfo.bEnableShadow = true;
+void SpatialDebuggerSystem::ActorAuthorityIntentChanged(Worker_EntityId EntityId, VirtualWorkerId NewIntentVirtualWorkerId) const
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:324]: The layout in this cpp file mixes SpatialDebugger and SpatialDebuggerSystem. Ideally each class has it's own cpp file.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h [hunk 14]
```diff
@@ -35,7 +36,44 @@ DECLARE_CYCLE_STAT(TEXT("Projection"), STAT_Projection, STATGROUP_SpatialDebugge
...
+
+	void ActorAuthorityIntentChanged(Worker_EntityId EntityId, VirtualWorkerId NewIntentVirtualWorkerId) const;
+
+	DECLARE_MULTICAST_DELEGATE_OneParam(FSpatialDebuggerActorAddedDelegate, AActor*);
+	FSpatialDebuggerActorAddedDelegate OnEntityActorAddedDelegate;
+
+	TOptional<SpatialDebugging> GetDebuggingData(Worker_EntityId Entity) const;
+	AActor* GetActor(Worker_EntityId EntityId) const;
+	const Worker_EntityId_Key* GetActorEntityId(AActor* Actor) const;
+	const TMap<Worker_EntityId_Key, TWeakObjectPtr<AActor>>& GetActors() const;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:60]: Worth typedefing this container type

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 12 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2807]: Yeah I tried that - would mix SpatialDebuggerSystem as a field name and as a class name. I can go for

```
using namespace SpatialGDK;

this->SpaitalDebuggerSystem = MakeUnique<SpatialGDK::SpatialDebuggerSystem>(...);
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 13 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:324]: Should I separate headers as well then?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 12 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2807]: using namespace SpatialGDK::FSubView etc. ?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 13 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:324]: Yes if it's not much hassle!

**Review by m-samiec — APPROVED** [review]
